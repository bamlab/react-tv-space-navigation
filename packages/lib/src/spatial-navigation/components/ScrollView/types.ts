export type CustomScrollViewRef = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- Already undocumented in RN https://github.com/facebook/react-native/blob/1c1c8335db2494672cf955cf4db574e23fd2198a/packages/react-native/Libraries/Components/ScrollView/ScrollView.d.ts#L861
  getInnerViewNode: () => any;
  scrollTo: (args: { x?: number; y?: number; animated: boolean }) => void;
};
