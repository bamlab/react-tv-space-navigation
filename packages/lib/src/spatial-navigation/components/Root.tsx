import { ReactNode, useRef } from 'react';
import { ParentIdContext } from '../context/ParentIdContext';
import { SpatialNavigatorContext } from '../context/SpatialNavigatorContext';
import { useBeforeMountEffect } from '../hooks/useBeforeMountEffect';
import { useCreateSpatialNavigator } from '../hooks/useCreateSpatialNavigator';
import { useRemoteControl } from '../hooks/useRemoteControl';
import { OnDirectionHandledWithoutMovement } from '../SpatialNavigator';

const ROOT_ID = 'root';

type Props = {
  /**
   * Determines if the spatial navigation is active.
   * If false, the spatial navigation will be locked, and no nodes can be focused.
   * This is useful to handle a multi page app: you can disable the non-focused pages' spatial navigation roots.
   */
  isActive?: boolean;
  /**
   * Called when you're reaching a border of the navigator.
   * A use case for this would be the implementation of a side menu
   * that's shared between pages. You can have a separate navigator
   * for your side menu, which would be common across pages, and you'd
   * make this menu active when you reach the left side of your page navigator.
   */
  onDirectionHandledWithoutMovement?: OnDirectionHandledWithoutMovement;
  children: ReactNode;
};

export const SpatialNavigationRoot = ({
  isActive = true,
  onDirectionHandledWithoutMovement = () => undefined,
  children,
}: Props) => {
  // We can't follow the react philosophy here: we can't recreate a navigator if this function changes
  // so we'll have to store its ref and update the ref if there is a new value to this function
  const onDirectionHandledWithoutMovementRef = useRef<OnDirectionHandledWithoutMovement>(
    () => undefined,
  );
  // Update the ref at every render
  onDirectionHandledWithoutMovementRef.current = onDirectionHandledWithoutMovement;

  const spatialNavigator = useCreateSpatialNavigator({
    onDirectionHandledWithoutMovementRef,
  });

  useRemoteControl({ spatialNavigator, isActive });

  useBeforeMountEffect(() => {
    spatialNavigator.registerNode(ROOT_ID, { orientation: 'vertical' });
    return () => spatialNavigator.unregisterNode(ROOT_ID);
  }, []);

  return (
    <SpatialNavigatorContext.Provider value={spatialNavigator}>
      <ParentIdContext.Provider value={ROOT_ID}>{children}</ParentIdContext.Provider>
    </SpatialNavigatorContext.Provider>
  );
};
