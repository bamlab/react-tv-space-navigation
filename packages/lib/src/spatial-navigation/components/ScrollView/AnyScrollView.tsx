import React from 'react';
import { ViewStyle, ScrollView } from 'react-native';
import { CustomScrollView } from './CustomScrollView/CustomScrollView';
import { CustomScrollViewRef } from './types';

type Props = {
  useNativeScroll: boolean;

  horizontal?: boolean;
  children: React.ReactNode;
  style?: ViewStyle;
  contentContainerStyle?: ViewStyle;
  scrollDuration?: number;
  onScroll?: (event: { nativeEvent: { contentOffset: { y: number; x: number } } }) => void;
  testID?: string;
};

export const AnyScrollView = React.forwardRef<CustomScrollViewRef, Props>(
  ({ useNativeScroll, ...props }: Props, ref) => {
    if (useNativeScroll) {
      return (
        <ScrollView
          ref={ref as React.RefObject<ScrollView>}
          showsHorizontalScrollIndicator={false}
          showsVerticalScrollIndicator={false}
          scrollEnabled={false}
          scrollEventThrottle={16}
          {...props}
        />
      );
    }

    return <CustomScrollView ref={ref} {...props} />;
  },
);

AnyScrollView.displayName = 'AnyScrollView';
