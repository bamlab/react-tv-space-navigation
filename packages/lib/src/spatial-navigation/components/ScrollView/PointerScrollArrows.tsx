import React, { ReactElement, ReactNode } from 'react';
import { View, ViewStyle, StyleSheet } from 'react-native';

export const PointerScrollArrows = React.memo(
  ({
    ascendingArrow,
    descendingArrowProps,
    ascendingArrowContainerStyle,
    descendingArrow,
    ascendingArrowProps,
    descendingArrowContainerStyle,
  }: {
    ascendingArrow?: ReactElement;
    ascendingArrowProps?: {
      onMouseEnter: () => void;
      onMouseLeave: () => void;
    };
    ascendingArrowContainerStyle?: ViewStyle;
    descendingArrow?: ReactNode;
    descendingArrowProps?: {
      onMouseEnter: () => void;
      onMouseLeave: () => void;
    };
    descendingArrowContainerStyle?: ViewStyle;
  }) => {
    return (
      <>
        <View
          style={[styles.arrowContainer, descendingArrowContainerStyle]}
          {...descendingArrowProps}
        >
          {descendingArrow}
        </View>
        <View style={ascendingArrowContainerStyle} {...ascendingArrowProps}>
          {ascendingArrow}
        </View>
      </>
    );
  },
);
PointerScrollArrows.displayName = 'PointerScrollArrows';

const styles = StyleSheet.create({
  arrowContainer: {
    position: 'absolute',
  },
});
