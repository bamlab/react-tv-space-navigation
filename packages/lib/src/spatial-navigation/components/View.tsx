import { forwardRef } from 'react';
import { StyleSheet, View, ViewStyle } from 'react-native';
import { NodeOrientation } from '../types/orientation';
import { SpatialNavigationNodeRef } from '../types/SpatialNavigationNodeRef';
import { SpatialNavigationNode } from './Node';

type Props = {
  children: React.ReactNode;
  style?: ViewStyle;
  direction: NodeOrientation;
  alignInGrid?: boolean;
};

export const SpatialNavigationView = forwardRef<SpatialNavigationNodeRef, Props>(
  ({ direction = 'horizontal', alignInGrid = false, children, style }: Props, ref) => {
    return (
      <SpatialNavigationNode orientation={direction} alignInGrid={alignInGrid} ref={ref}>
        <View
          style={[style, stylesByDirection[direction]]}
        >
          {children}
        </View>
      </SpatialNavigationNode>
    );
  },
);
SpatialNavigationView.displayName = 'SpatialNavigationView';

const styles = StyleSheet.create({
  viewVertical: {
    display: 'flex',
    flexDirection: 'column',
  },
  viewHorizontal: {
    display: 'flex',
    flexDirection: 'row',
  },
  viewVerticalReverse: {
    display: 'flex',
    flexDirection: 'column-reverse',
  },
  viewHorizontalReverse: {
    display: 'flex',
    flexDirection: 'row-reverse',
  },
});

const stylesByDirection = {
  'horizontal':  styles.viewHorizontal,
  'vertical':  styles.viewVertical,
  'horizontal-reverse':  styles.viewHorizontalReverse,
  'vertical-reverse':  styles.viewVerticalReverse,
} as const;