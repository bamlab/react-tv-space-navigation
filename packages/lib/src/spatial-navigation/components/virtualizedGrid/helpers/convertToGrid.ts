import chunk from 'lodash/chunk';
import { GridRowType } from '../SpatialNavigationVirtualizedGrid';
import { ItemWithIndex } from '../../virtualizedList/VirtualizedList';
import { NodeOrientation } from '../../../types/orientation';

export const convertToGrid = <T extends ItemWithIndex>(
  data: T[],
  numberOfColumns: number,
  header?: JSX.Element,
): GridRowType<T>[] => {
  const rows: T[][] = chunk(data, numberOfColumns);

  return rows.map((items, index) => {
    //We do this to have index taking into account the header
    const computedIndex = header ? index + 1 : index;
    return { items, index: computedIndex };
  });
};

export const invertOrientation = (orientation: NodeOrientation): NodeOrientation =>
  orientation === 'vertical' ? 'horizontal' : 'vertical';
