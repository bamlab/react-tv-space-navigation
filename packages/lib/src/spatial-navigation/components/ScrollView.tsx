import React, { useCallback, RefObject, useRef, ReactElement, ReactNode, useMemo } from 'react';
import {
  ScrollView,
  View,
  ViewStyle,
  StyleSheet,
  Platform,
  NativeSyntheticEvent,
  NativeScrollEvent,
} from 'react-native';
import {
  SpatialNavigatorParentScrollContext,
  useSpatialNavigatorParentScroll,
} from '../context/ParentScrollContext';
import { scrollToNewlyFocusedElement } from '../helpers/scrollToNewlyfocusedElement';
import { useSpatialNavigationDeviceType } from '../context/DeviceContext';

type Props = {
  horizontal?: boolean;
  /**
   * Use this offset to prevent the element from sticking too closely to the edges of the screen during scrolling.
   * This is a margin in pixels.
   */
  offsetFromStart?: number;
  children: React.ReactNode;
  style?: ViewStyle;
  /** Arrow that will show up inside the arrowContainer */
  descendingArrow?: ReactElement;
  /** Arrow that will show up inside the arrowContainer */
  ascendingArrow?: ReactElement;
  /** Style props for the arrow container, basically the area hoverable that triggers a scroll  */
  descendingArrowContainerStyle?: ViewStyle;
  /** Style props for the arrow container, basically the area hoverable that triggers a scroll  */
  ascendingArrowContainerStyle?: ViewStyle;
  /** Number of pixels scrolled every 10ms - only when using web cursor pointer to scroll */
  pointerScrollSpeed?: number;
};

const useRemotePointerScrollviewScrollProps = ({
  pointerScrollSpeed,
  scrollY,
  scrollViewRef,
}: {
  pointerScrollSpeed: number;
  scrollY: React.MutableRefObject<number>;
  scrollViewRef: React.MutableRefObject<ScrollView | null>;
}) => {
  const {
    deviceType,
    getScrollingIntervalId: getScrollingId,
    setScrollingIntervalId: setScrollingId,
  } = useSpatialNavigationDeviceType();

  const onMouseEnterTop = useCallback(() => {
    if (deviceType === 'remotePointer') {
      let currentScrollPosition = scrollY.current;
      const id = setInterval(() => {
        currentScrollPosition -= pointerScrollSpeed;
        scrollViewRef.current?.scrollTo({
          y: currentScrollPosition,
          animated: false,
        });
      }, 10);
      setScrollingId(id);
    }
  }, [deviceType, pointerScrollSpeed, scrollY, scrollViewRef, setScrollingId]);

  const onMouseEnterBottom = useCallback(() => {
    if (deviceType === 'remotePointer') {
      let currentScrollPosition = scrollY.current;
      const id = setInterval(() => {
        currentScrollPosition += pointerScrollSpeed;
        scrollViewRef.current?.scrollTo({
          y: currentScrollPosition,
          animated: false,
        });
      }, 10);
      setScrollingId(id);
    }
  }, [deviceType, pointerScrollSpeed, scrollY, scrollViewRef, setScrollingId]);

  const onMouseLeave = useCallback(() => {
    if (deviceType === 'remotePointer') {
      const intervalId = getScrollingId();
      if (intervalId) {
        clearInterval(intervalId);
        setScrollingId(null);
      }
    }
  }, [deviceType, getScrollingId, setScrollingId]);

  const ascendingArrowProps = useMemo(
    () =>
      Platform.select({
        web: { onMouseEnter: onMouseEnterBottom, onMouseLeave: onMouseLeave },
      }),
    [onMouseEnterBottom, onMouseLeave],
  );

  const descendingArrowProps = useMemo(
    () =>
      Platform.select({
        web: { onMouseEnter: onMouseEnterTop, onMouseLeave: onMouseLeave },
      }),
    [onMouseEnterTop, onMouseLeave],
  );

  return {
    deviceType,
    ascendingArrowProps,
    descendingArrowProps,
  };
};

export const SpatialNavigationScrollView = ({
  horizontal = false,
  style,
  offsetFromStart = 0,
  children,
  ascendingArrow,
  ascendingArrowContainerStyle,
  descendingArrow,
  descendingArrowContainerStyle,
  pointerScrollSpeed = 10,
}: Props) => {
  const { scrollToNodeIfNeeded: makeParentsScrollToNodeIfNeeded } =
    useSpatialNavigatorParentScroll();
  const scrollViewRef = useRef<ScrollView>(null);

  const scrollY = useRef<number>(0);

  const { ascendingArrowProps, descendingArrowProps, deviceType } =
    useRemotePointerScrollviewScrollProps({ pointerScrollSpeed, scrollY, scrollViewRef });

  const scrollToNode = useCallback(
    (newlyFocusedElementRef: RefObject<View>) => {
      try {
        if (deviceType === 'remoteKeys') {
          newlyFocusedElementRef?.current?.measureLayout(
            scrollViewRef?.current?.getInnerViewNode(),
            (left, top) =>
              scrollToNewlyFocusedElement({
                newlyFocusedElementDistanceToLeftRelativeToLayout: left,
                newlyFocusedElementDistanceToTopRelativeToLayout: top,
                horizontal,
                offsetFromStart,
                scrollViewRef,
              }),
            () => {},
          );
        }
      } catch {
        // A crash can happen when calling measureLayout when a page unmounts. No impact on focus detected in regular use cases.
      }
      makeParentsScrollToNodeIfNeeded(newlyFocusedElementRef); // We need to propagate the scroll event for parents if we have nested ScrollViews/VirtualizedLists.
    },
    [makeParentsScrollToNodeIfNeeded, horizontal, offsetFromStart, deviceType],
  );

  const onScroll = useCallback(
    (event: NativeSyntheticEvent<NativeScrollEvent>) => {
      scrollY.current = event.nativeEvent.contentOffset.y;
    },
    [scrollY],
  );

  return (
    <SpatialNavigatorParentScrollContext.Provider value={scrollToNode}>
      <ScrollView
        ref={scrollViewRef}
        horizontal={horizontal}
        style={style}
        showsHorizontalScrollIndicator={false}
        showsVerticalScrollIndicator={false}
        scrollEnabled={false}
        onScroll={onScroll}
        scrollEventThrottle={16}
      >
        {children}
      </ScrollView>
      {deviceType === 'remotePointer' ? (
        <PointerScrollArrows
          descendingArrow={descendingArrow}
          ascendingArrow={ascendingArrow}
          descendingArrowContainerStyle={descendingArrowContainerStyle}
          ascendingArrowContainerStyle={ascendingArrowContainerStyle}
          ascendingArrowProps={ascendingArrowProps}
          descendingArrowProps={descendingArrowProps}
        />
      ) : undefined}
    </SpatialNavigatorParentScrollContext.Provider>
  );
};

const PointerScrollArrows = React.memo(
  ({
    ascendingArrow,
    descendingArrowProps,
    ascendingArrowContainerStyle,
    descendingArrow,
    ascendingArrowProps,
    descendingArrowContainerStyle,
  }: {
    ascendingArrow?: ReactElement;
    ascendingArrowProps?: {
      onMouseEnter: () => void;
      onMouseLeave: () => void;
    };
    ascendingArrowContainerStyle?: ViewStyle;
    descendingArrow?: ReactNode;
    descendingArrowProps?: {
      onMouseEnter: () => void;
      onMouseLeave: () => void;
    };
    descendingArrowContainerStyle?: ViewStyle;
  }) => {
    return (
      <>
        <View
          style={[styles.arrowContainer, descendingArrowContainerStyle]}
          {...descendingArrowProps}
        >
          {descendingArrow}
        </View>
        <View style={ascendingArrowContainerStyle} {...ascendingArrowProps}>
          {ascendingArrow}
        </View>
      </>
    );
  },
);
PointerScrollArrows.displayName = 'PointerScrollArrows';

const styles = StyleSheet.create({
  arrowContainer: {
    position: 'absolute',
  },
});
