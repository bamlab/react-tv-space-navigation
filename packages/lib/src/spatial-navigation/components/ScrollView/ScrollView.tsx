import React, { useCallback, RefObject, useRef, ReactElement, forwardRef } from 'react';
import { ScrollView, View, ViewStyle, Platform } from 'react-native';
import {
  SpatialNavigatorParentScrollContext,
  useSpatialNavigatorParentScroll,
} from '../../context/ParentScrollContext';
import { scrollToNewlyFocusedElement } from '../../helpers/scrollToNewlyfocusedElement';
import { mergeRefs } from '../../helpers/mergeRefs';
import { useRemotePointerScrollviewScrollProps } from './pointer/useRemotePointerScrollviewScrollProps';
import { PointerScrollArrows } from './pointer/PointerScrollArrows';
import { CustomScrollViewRef } from './types';
import { AnyScrollView } from './AnyScrollView';

type Props = {
  horizontal?: boolean;
  /**
   * Use this offset to prevent the element from sticking too closely to the edges of the screen during scrolling.
   * This is a margin in pixels.
   */
  offsetFromStart?: number;
  children: React.ReactNode;
  style?: ViewStyle;
  contentContainerStyle?: ViewStyle;
  /** Arrow that will show up inside the arrowContainer */
  descendingArrow?: ReactElement;
  /** Arrow that will show up inside the arrowContainer */
  ascendingArrow?: ReactElement;
  /** Style props for the arrow container, basically the area hoverable that triggers a scroll  */
  descendingArrowContainerStyle?: ViewStyle;
  /** Style props for the arrow container, basically the area hoverable that triggers a scroll  */
  ascendingArrowContainerStyle?: ViewStyle;
  /** Number of pixels scrolled every 10ms - only when using web cursor pointer to scroll */
  pointerScrollSpeed?: number;
  /** Toggles the native scrolling version of the scroll view instead of the CSS scroll */
  useNativeScroll?: boolean;
  /** Configures the scroll duration in the case of CSS scroll */
  scrollDuration?: number;
  testID?: string;
};

const getNodeRef = (node: CustomScrollViewRef | null | undefined) => {
  if (Platform.OS === 'web') {
    return node?.getInnerViewNode();
  }

  return node;
};

export const SpatialNavigationScrollView = forwardRef<ScrollView, Props>(
  (
    {
      horizontal = false,
      style,
      offsetFromStart = 0,
      children,
      ascendingArrow,
      ascendingArrowContainerStyle,
      descendingArrow,
      descendingArrowContainerStyle,
      pointerScrollSpeed = 10,
      contentContainerStyle,
      useNativeScroll = false,
      scrollDuration = 200,
      testID,
    },
    ref,
  ) => {
    const { scrollToNodeIfNeeded: makeParentsScrollToNodeIfNeeded } =
      useSpatialNavigatorParentScroll();
    const scrollViewRef = useRef<CustomScrollViewRef>(null);

    const scrollY = useRef<number>(0);

    const { ascendingArrowProps, descendingArrowProps, deviceType, deviceTypeRef } =
      useRemotePointerScrollviewScrollProps({ pointerScrollSpeed, scrollY, scrollViewRef });

    const scrollToNode = useCallback(
      (newlyFocusedElementRef: RefObject<View>, additionalOffset = 0) => {
        try {
          if (deviceTypeRef.current === 'remoteKeys') {
            newlyFocusedElementRef?.current?.measureLayout(
              getNodeRef(scrollViewRef?.current),
              (left, top) =>
                scrollToNewlyFocusedElement({
                  newlyFocusedElementDistanceToLeftRelativeToLayout: left,
                  newlyFocusedElementDistanceToTopRelativeToLayout: top,
                  horizontal,
                  offsetFromStart: offsetFromStart + additionalOffset,
                  scrollViewRef,
                }),
              () => {},
            );
          }
        } catch {
          // A crash can happen when calling measureLayout when a page unmounts. No impact on focus detected in regular use cases.
        }
        makeParentsScrollToNodeIfNeeded(newlyFocusedElementRef, additionalOffset); // We need to propagate the scroll event for parents if we have nested ScrollViews/VirtualizedLists.
      },
      [makeParentsScrollToNodeIfNeeded, horizontal, offsetFromStart, deviceTypeRef],
    );

    const onScroll = useCallback(
      (event: { nativeEvent: { contentOffset: { y: number; x: number } } }) => {
        scrollY.current = event.nativeEvent.contentOffset.y;
      },
      [scrollY],
    );

    return (
      <SpatialNavigatorParentScrollContext.Provider value={scrollToNode}>
        <AnyScrollView
          useNativeScroll={useNativeScroll}
          scrollDuration={scrollDuration}
          ref={mergeRefs([ref, scrollViewRef])}
          horizontal={horizontal}
          style={style}
          contentContainerStyle={contentContainerStyle}
          onScroll={onScroll}
          testID={testID}
        >
          {children}
        </AnyScrollView>
        {deviceType === 'remotePointer' ? (
          <PointerScrollArrows
            descendingArrow={descendingArrow}
            ascendingArrow={ascendingArrow}
            descendingArrowContainerStyle={descendingArrowContainerStyle}
            ascendingArrowContainerStyle={ascendingArrowContainerStyle}
            ascendingArrowProps={ascendingArrowProps}
            descendingArrowProps={descendingArrowProps}
          />
        ) : undefined}
      </SpatialNavigatorParentScrollContext.Provider>
    );
  },
);
SpatialNavigationScrollView.displayName = 'SpatialNavigationScrollView';
