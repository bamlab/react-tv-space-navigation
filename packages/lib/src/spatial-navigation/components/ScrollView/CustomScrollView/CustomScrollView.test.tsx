import { View } from 'react-native';
import { RenderResult, act, fireEvent, render, screen } from '@testing-library/react-native';
import { ReactTestInstance } from 'react-test-renderer';
import '../../tests/helpers/configureTestRemoteControl';
import { SpatialNavigationScrollView } from '../ScrollView';
import { SpatialNavigationView } from '../../View';
import { TestButton } from '../../tests/TestButton';
import { SpatialNavigationRoot } from '../../Root';
import testRemoteControlManager from '../../tests/helpers/testRemoteControlManager';
import { DefaultFocus } from '../../../context/DefaultFocusContext';

const MOCKED_BUTTON_HEIGHT = 100;

export const setComponentLayoutSize = (
  component: ReactTestInstance,
  size: { width: number; height: number; x: number; y: number },
) => {
  fireEvent(component, 'layout', {
    nativeEvent: { layout: { width: size.width, height: size.height, x: size.x, y: size.y } },
  });
};

export const expectButtonToHaveFocus = (component: RenderResult, text: string) => {
  const element = component.getByRole('button', { name: text });
  expect(element).toBeSelected();
};

const TestPage = () => {
  return (
    <SpatialNavigationRoot>
      <DefaultFocus>
        <SpatialNavigationScrollView testID="scrollview">
          <SpatialNavigationView direction="vertical">
            <TestButton title="1" />
            <TestButton title="2" />
            <TestButton title="3" />
            <TestButton title="4" />
            <TestButton title="5" />
            <TestButton title="6" />
            <TestButton title="7" />
            <TestButton title="8" />
          </SpatialNavigationView>
        </SpatialNavigationScrollView>
      </DefaultFocus>
    </SpatialNavigationRoot>
  );
};

jest.spyOn(View.prototype, 'measureLayout').mockImplementation(function (node, callback) {
  // @ts-expect-error it's weird but that's fine, it's our only way to get the button's context
  const buttonLabel = this?.props?.accessibilityLabel;
  const buttonNumber = parseInt(buttonLabel, 10) - 1;

  callback(0, buttonNumber * MOCKED_BUTTON_HEIGHT, 0, 0);
});

const expectViewToHaveScroll = (element: ReactTestInstance, scrollValue: number) =>
  expect(element).toHaveStyle({ transform: [{ translateY: scrollValue }] });

describe('CustomScrollView', () => {
  const scrollViewTestId = 'scrollview';
  const innerScrollViewTestId = scrollViewTestId + '-content';

  it('scrolls properly upon focus and stops when overflowing', async () => {
    const MOCK_SCREEN_SIZE = 300;
    const MOCK_TOTAL_CONTENT_SIZE = 800;

    const component = render(<TestPage />);
    act(() => jest.runAllTimers());

    const scrollViewRoot = component.getByTestId(scrollViewTestId);
    setComponentLayoutSize(scrollViewRoot, { width: 0, height: MOCK_SCREEN_SIZE, x: 0, y: 0 });
    const scrollViewInner = component.getByTestId(innerScrollViewTestId);
    setComponentLayoutSize(scrollViewInner, {
      width: 0,
      height: MOCK_TOTAL_CONTENT_SIZE,
      x: 0,
      y: 0,
    });

    testRemoteControlManager.handleDown();
    expectViewToHaveScroll(scrollViewInner, -100);
    testRemoteControlManager.handleDown();
    expectViewToHaveScroll(scrollViewInner, -200);
    testRemoteControlManager.handleDown();
    expectViewToHaveScroll(scrollViewInner, -300);
    testRemoteControlManager.handleDown();
    expectViewToHaveScroll(scrollViewInner, -400);
    testRemoteControlManager.handleDown();
    expectViewToHaveScroll(scrollViewInner, -500);
    testRemoteControlManager.handleDown();

    // Once the view is going to over-scroll, it should stop scrolling
    expectViewToHaveScroll(scrollViewInner, -500);
    testRemoteControlManager.handleDown();
    expectViewToHaveScroll(scrollViewInner, -500);
    testRemoteControlManager.handleDown();
    expectViewToHaveScroll(scrollViewInner, -500);
    testRemoteControlManager.handleDown();

    expect(screen).toMatchSnapshot();
  });
});
