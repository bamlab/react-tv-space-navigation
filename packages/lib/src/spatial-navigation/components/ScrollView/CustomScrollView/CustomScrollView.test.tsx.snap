// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomScrollView scrolls properly upon focus and stops when overflowing 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "flex": 1,
        "flexDirection": "column",
        "overflow": "hidden",
      },
      undefined,
    ]
  }
  testID="scrollview"
>
  <View
    collapsable={false}
    onLayout={[Function]}
    style={
      {
        "transform": [
          {
            "translateY": -500,
          },
        ],
      }
    }
    testID="scrollview-content"
  >
    <View
      style={
        [
          undefined,
          {
            "display": "flex",
            "flexDirection": "column",
          },
        ]
      }
    >
      <View
        accessibilityHint="1"
        accessibilityLabel="1"
        accessibilityRole="button"
        accessibilityState={
          {
            "selected": false,
          }
        }
        accessible={true}
        isFocused={false}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderRadius": 100,
              "padding": 6,
            },
            undefined,
          ]
        }
      >
        <Text>
          1
        </Text>
      </View>
      <View
        accessibilityHint="2"
        accessibilityLabel="2"
        accessibilityRole="button"
        accessibilityState={
          {
            "selected": false,
          }
        }
        accessible={true}
        isFocused={false}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderRadius": 100,
              "padding": 6,
            },
            undefined,
          ]
        }
      >
        <Text>
          2
        </Text>
      </View>
      <View
        accessibilityHint="3"
        accessibilityLabel="3"
        accessibilityRole="button"
        accessibilityState={
          {
            "selected": false,
          }
        }
        accessible={true}
        isFocused={false}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderRadius": 100,
              "padding": 6,
            },
            undefined,
          ]
        }
      >
        <Text>
          3
        </Text>
      </View>
      <View
        accessibilityHint="4"
        accessibilityLabel="4"
        accessibilityRole="button"
        accessibilityState={
          {
            "selected": false,
          }
        }
        accessible={true}
        isFocused={false}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderRadius": 100,
              "padding": 6,
            },
            undefined,
          ]
        }
      >
        <Text>
          4
        </Text>
      </View>
      <View
        accessibilityHint="5"
        accessibilityLabel="5"
        accessibilityRole="button"
        accessibilityState={
          {
            "selected": false,
          }
        }
        accessible={true}
        isFocused={false}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderRadius": 100,
              "padding": 6,
            },
            undefined,
          ]
        }
      >
        <Text>
          5
        </Text>
      </View>
      <View
        accessibilityHint="6"
        accessibilityLabel="6"
        accessibilityRole="button"
        accessibilityState={
          {
            "selected": false,
          }
        }
        accessible={true}
        isFocused={false}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderRadius": 100,
              "padding": 6,
            },
            undefined,
          ]
        }
      >
        <Text>
          6
        </Text>
      </View>
      <View
        accessibilityHint="7"
        accessibilityLabel="7"
        accessibilityRole="button"
        accessibilityState={
          {
            "selected": false,
          }
        }
        accessible={true}
        isFocused={false}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderRadius": 100,
              "padding": 6,
            },
            undefined,
          ]
        }
      >
        <Text>
          7
        </Text>
      </View>
      <View
        accessibilityHint="8"
        accessibilityLabel="8"
        accessibilityRole="button"
        accessibilityState={
          {
            "selected": true,
          }
        }
        accessible={true}
        isFocused={true}
        style={
          [
            {
              "backgroundColor": "red",
              "borderRadius": 100,
              "padding": 6,
            },
            undefined,
          ]
        }
      >
        <Text>
          8
        </Text>
      </View>
    </View>
  </View>
</View>
`;
